### PR TITLE
Upgrade to new metadata format

### PR DESCRIPTION
**Pull request by fjammes**
- Implement new metadata format for cosmoDC2
- Improve ingest/file parameters management
- Add unit tests
- Manage http load balancing for input data

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -9,7 +9,7 @@ before_script:
   - if [ "$DOCKER_USERNAME" ]; then docker login -u="$DOCKER_USERNAME" -p="$DOCKER_PASSWORD"; fi
   - git clone --depth 1 -b "k8s-v1.18.2" --single-branch https://github.com/k8s-school/kind-travis-ci.git
   - ./kind-travis-ci/k8s-create.sh -s
-  - git clone --depth 1 -b "master" --single-branch https://github.com/lsst/qserv-operator
+  - git clone --depth 1 -b "tickets/DM-28108" --single-branch https://github.com/lsst/qserv-operator
```
Comment (iagaponenko) [inline .travis.yml]: Are you going to keep the specific branch name in this file? Is this by intent?

File: example/query.sh [hunk 2]
```diff
@@ -10,6 +10,7 @@ CZAR_HOST="${INSTANCE}-czar-0.${INSTANCE}-czar"
 DATABASE="dc2_run2_1i_dr1b"
```
Comment (iagaponenko) [inline example/query.sh:10]: Don't you need to comment this out (even though it's overwritten later)?

File: manifests/base/dev/kustomization.yaml [hunk 3]
```diff
@@ -7,8 +7,10 @@ resources:
 
 configMapGenerator:
 - name: config-data-url
+  files:
+  - servers.json
   literals:
-  - DATA_URL="https://raw.githubusercontent.com/lsst-dm/qserv-ingest/master/data/example_db/"
+  - DATA_URL="https://raw.githubusercontent.com/lsst-dm/qserv-ingest/tickets/DM-28108/data/cosmoDC2_v1_1_4_image_testintegration/"
```
Comment (iagaponenko) [inline manifests/base/dev/kustomization.yaml]: Should this be a specific branch?

File: manifests/base/init/kustomization.yaml [hunk 4]
```diff
@@ -9,16 +9,14 @@ resources:
...
-  disableNameSuffixHash: true
+  - qserv
 
 configMapGenerator:
 - name: config-data-url
+  files:
+  - servers.json
   literals:
-  - DATA_URL="https://raw.githubusercontent.com/lsst-dm/qserv-ingest/master/data/example_db/"
+  - DATA_URL="https://raw.githubusercontent.com/lsst-dm/qserv-ingest/tickets/DM-28108/data/cosmoDC2_v1_1_4_image_testintegration/"
```
Comment (iagaponenko) [inline manifests/base/init/kustomization.yaml]: Should this be a specific branch?

File: manifests/base/init/servers.json [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
+    "http_servers":[
+        "https://raw.githubusercontent.com/lsst-dm/qserv-ingest/tickets/DM-28108/data/cosmoDC2_v1_1_4_image_testintegration/",
+        "https://raw.githubusercontent.com/lsst-dm/qserv-ingest/tickets/DM-28108/data/cosmoDC2_v1_1_4_image_testintegration/"
+    ]
```
Comment (iagaponenko) [inline manifests/base/init/servers.json:5]: Should this be a specific branch?

File: rootfs/ingest/bin/replctl-register [hunk 6]
```diff
@@ -94,7 +94,7 @@ if __name__ == "__main__":
     url = urllib.parse.urljoin(args.url, path)
     json = {"database": metadata.database,
             "CAINFO": "/etc/pki/tls/certs/ca-bundle.crt",
-            "SSL_VERIFYPEER": 0
+            "SSL_VERIFYPEER": 1
```
Comment (iagaponenko) [inline rootfs/ingest/bin/replctl-register:97]: That won't be need since the default behavior of the Ingest service is to verify peers. Though, you may still keep this option here for the sake of clarity.

File: rootfs/ingest/python/qserv/ingest.py [hunk 7]
```diff
@@ -210,49 +251,42 @@ def _get_chunk_location(repl_url, chunk, database, transaction_id):
 
     # Get location host and port
     host = responseJson["location"]["host"]
-    port = responseJson["location"]["port"]
+    # FIXME Ask Igor Gaponenko for a method to get this parameter?
+    # Or set it from qserv-operator?
+    port = 25004
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/ingest.py]: The documentation on this service https://confluence.lsstcorp.org/pages/viewpage.action?pageId=135955129 says that you should use:
```
host = responseJson["location"]["http_host"]
port = responseJson["location"]["http_port"]
```

File: rootfs/ingest/python/qserv/ingest.py [hunk 8]
```diff
@@ -210,49 +251,42 @@ def _get_chunk_location(repl_url, chunk, database, transaction_id):
...
 
-def _ingest_chunk(args):
-    repl_url, chunk_meta, transaction_id, chunk = args
+def _ingest_chunk(ingest_args):
+
+    kwargs = ingest_args.get_kwargs()
+    chunk_file_url = kwargs['chunk_file_url']
     try:
-        logging.debug("Start INGEST")
+        logging.debug("Start ingest of chunk file: %s", chunk_file_url)
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/ingest.py]: I would rephrase this:
```
Start ingesting chunk contribution:
```

File: rootfs/ingest/python/qserv/ingest.py [hunk 9]
```diff
@@ -210,49 +251,42 @@ def _get_chunk_location(repl_url, chunk, database, transaction_id):
...
-        if chunk_meta.is_director(table):
-            _ingest_file(host, port, transaction_id, chunk_base_url, chunk_id, table, True)
+
+        _ingest_file(**kwargs)
         endedAt = time.strftime("%H:%M:%S", time.localtime())
     except Exception as e:
-        _LOG.critical('Error in ingest task for chunk %s: %s',
-                        chunk,
-                        e)
+        _LOG.critical('Error while ingesting chunk file %s: %s',
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/ingest.py]:
```
Error while ingesting chunk contribution:
```

File: rootfs/ingest/python/qserv/test_util.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    """Check if a file exists on a remote HTTP server
+    """
+    assert qserv.util.http_file_exists("https://www.k8s-school.fr/team/", "index.html")
+    assert not qserv.util.http_file_exists("https://www.k8s-school.fr/team/", "false.html")
+    
+def test_json_get():
+    
+    data = qserv.util.json_get(__file__,"servers.json")
+    assert (data['http_servers'][0] == "https://server1")
+    assert (data['http_servers'][2] == "https://server3")
```
Comment (iagaponenko) [inline rootfs/ingest/python/qserv/test_util.py]: No newline in the end of the file

**Review by iagaponenko — APPROVED** [review]
Looks good to me.
I have also left a few comments and suggestion on the code.

File: .travis.yml (hunk 1 above)
Comment (fjammes) [reply .travis.yml]: Yes I need to change it when merging to master!

File: example/query.sh (hunk 2 above)
Comment (fjammes) [reply example/query.sh:10]: Here, I need to make a cleaner test procedure in a future ticket.

File: manifests/base/dev/kustomization.yaml (hunk 3 above)
Comment (fjammes) [reply manifests/base/dev/kustomization.yaml]: I'll change it when merging thanks!

File: manifests/base/init/kustomization.yaml (hunk 4 above)
Comment (fjammes) [reply manifests/base/init/kustomization.yaml]: Idem than above

File: manifests/base/init/servers.json (hunk 5 above)
Comment (fjammes) [reply manifests/base/init/servers.json:5]: idem

File: rootfs/ingest/bin/replctl-register (hunk 6 above)
Comment (fjammes) [reply rootfs/ingest/bin/replctl-register:97]: :+1:

File: rootfs/ingest/python/qserv/ingest.py (hunk 7 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/ingest.py]: :+1:

File: rootfs/ingest/python/qserv/ingest.py (hunk 8 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/ingest.py]: :+1:

File: rootfs/ingest/python/qserv/ingest.py (hunk 9 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/ingest.py]: :+1:

File: rootfs/ingest/python/qserv/test_util.py (hunk 10 above)
Comment (fjammes) [reply rootfs/ingest/python/qserv/test_util.py]: :+1:
